rtv/mime_parsers.py

```diff
@@ -104,64 +104,40 @@ def get_mimetype(url):
 
 class ImgurMIMEParser(BaseMIMEParser):
     """
-    The majority of imgur links don't point directly to the image, so we need
-    to open the provided url and scrape the page for the link.
+    Imgur provides a json api exposing its entire infrastructure. Each imgur
+    page has an associated hash and can either contain an album, a gallery, or single image.
 
-    Scrape the actual image url from an imgur landing page. Imgur intentionally
-    obscures this on most reddit links in order to draw more traffic for their
-    advertisements.
-
-    There are a couple of <meta> tags that supply the relevant info:
-        <meta name="twitter:image" content="https://i.imgur.com/xrqQ4LEh.jpg">
-        <meta property="og:image" content="http://i.imgur.com/xrqQ4LE.jpg?fb">
-        <link rel="image_src" href="http://i.imgur.com/xrqQ4LE.jpg">
+    see https://apidocs.imgur.com
     """
     pattern = re.compile(r'https?://(w+\.)?(m\.)?imgur\.com/[^.]+$')
 
     @staticmethod
     def get_mimetype(url):
-        page = requests.get(url)
-        soup = BeautifulSoup(page.content, 'html.parser')
-        tag = soup.find('meta', attrs={'name': 'twitter:image'})
-        if tag:
-            url = tag.get('content')
-            if GifvMIMEParser.pattern.match(url):
-                return GifvMIMEParser.get_mimetype(url)
-        return BaseMIMEParser.get_mimetype(url)
-
-
-class ImgurAlbumMIMEParser(BaseMIMEParser):
-    """
-    Imgur albums can contain several images, which need to be scraped from the
-    landing page. Assumes the following html structure:
-
-        <div class="post-image">
-            <a href="//i.imgur.com/L3Lfp1O.jpg" class="zoom">
-                <img class="post-image-placeholder"
-                     src="//i.imgur.com/L3Lfp1Og.jpg" alt="Close up">
-                <img class="js-post-image-thumb"
-                     src="//i.imgur.com/L3Lfp1Og.jpg" alt="Close up">
-            </a>
-        </div>
-    """
-    pattern = re.compile(r'https?://(w+\.)?(m\.)?imgur\.com/a(lbum)?/[^.]+$')
-
-    @staticmethod
-    def get_mimetype(url):
-        page = requests.get(url)
-        soup = BeautifulSoup(page.content, 'html.parser')
-
-        urls = []
-        for div in soup.find_all('div', class_='post-image'):
-            img = div.find('img')
-            src = img.get('src') if img else None
-            if src:
-                urls.append('http:{0}'.format(src))
-
-        if urls:
-            return " ".join(urls), 'image/x-imgur-album'
-        else:
-            return url, None
+        endpoint = 'https://api.imgur.com/3/{domain}/{page_hash}'
+        header = {'authorization': 'Client-ID {}'.format('d8842d573e8b9dd')}
+
+        pattern = re.compile(r'https?://(w+\.)?(m\.)?imgur\.com/((?P<domain>a|album|gallery)/)?(?P<hash>.+)$')
+        m = pattern.match(url)
+        page_hash = m.group('hash')
+        domain = 'album' if m.group('domain') in ['a', 'album'] else 'gallery'
```

Comment:
There doesn't seem to be a way to reliably distinguish between gallery and image pages just from the url:

gallery: [https://imgur.com/NAVdRye](https://imgur.com/NAVdRye)
image: [https://imgur.com/0FiXBEw](https://imgur.com/0FiXBEw)

So I assume a given url is for a gallery page, which appear to be more common, and fallback to an image request upon failure.